pkg/accounting/accounting.go

```diff
@@ -0,0 +1,212 @@
+type Accounting struct {
+	balancesMu          sync.Mutex // mutex for accessing the balances map
+	balances            map[string]*PeerBalance
+	logger              logging.Logger
+	store               storage.StateStorer
+	disconnectThreshold uint64 // the debt threshold at which we will disconnect from a peer
+}
+
+// NewAccounting creates a new Accounting instance with the provided options
+func NewAccounting(o Options) *Accounting {
+	return &Accounting{
+		balances:            make(map[string]*PeerBalance),
+		disconnectThreshold: o.DisconnectThreshold,
+		logger:              o.Logger,
+		store:               o.Store,
+	}
+}
+
+// Reserve reserves a portion of the balance for peer
+func (a *Accounting) Reserve(peer swarm.Address, price uint64) error {
+	balance, err := a.getPeerBalance(peer)
+	if err != nil {
+		return err
+	}
+
+	balance.lock.Lock()
+	defer balance.lock.Unlock()
+
+	// the previously reserved balance plus the new price is the maximum amount paid if all current operations are successful
+	// since we pay this we have to reduce this (positive quantity) from the balance
+	// the disconnectThreshold is stored as a positive value which is why it must be negated prior to comparison
+	if balance.balance-int64(balance.reserved+price) < -int64(a.disconnectThreshold) {
+		return fmt.Errorf("cannot afford operation with peer %v", peer.String())
+	}
+
+	balance.reserved += price
+
+	return nil
+}
+
+// Release releases reserved funds
+func (a *Accounting) Release(peer swarm.Address, price uint64) {
+	balance, err := a.getPeerBalance(peer)
+	if err != nil {
+		a.logger.Error("attempting to release balance for peer that is not loaded")
+		return
+	}
+
+	balance.lock.Lock()
+	defer balance.lock.Unlock()
+
+	if price > balance.reserved {
+		// If Reserve and Release calls are always paired this should never happen
+		a.logger.Error("attempting to release more balance than was reserved for peer")
+		balance.reserved = 0
+	} else {
+		balance.reserved -= price
+	}
+}
+
+// Credit increases the balance the peer has with us
+func (a *Accounting) Credit(peer swarm.Address, price uint64) error {
+	balance, err := a.getPeerBalance(peer)
+	if err != nil {
+		return err
+	}
+
+	balance.lock.Lock()
+	defer balance.lock.Unlock()
+
+	nextBalance := balance.balance - int64(price)
+
+	a.logger.Tracef("crediting peer %v with price %d, new balance is %d", peer, price, nextBalance)
+
+	err = a.store.Put(a.balanceKey(peer), nextBalance)
+	if err != nil {
+		return err
+	}
+
+	balance.balance = nextBalance
+
+	// TODO: try to initiate payment if payment threshold is reached
+	// if balance.balance < -int64(a.paymentThreshold) { }
+
+	return nil
+}
+
+// Debit increases the balance we have with the peer
+func (a *Accounting) Debit(peer swarm.Address, price uint64) error {
+	balance, err := a.getPeerBalance(peer)
+	if err != nil {
+		return err
+	}
+
+	balance.lock.Lock()
+	defer balance.lock.Unlock()
+
+	nextBalance := balance.balance + int64(price)
```

Comment:
I would not do that change now. If there are `.debit` and `.credit` functions on `PeerBalance` they also should do the in-memory update of the balance (instead of returning `nextBalance`) which in the current code can lead to case where `Debit`/`Credit` returns a (non-disconnect) error but balances are still updated.

Reply:
Fair point, I did not take side effects into consideration.